### PR TITLE
Add docker transport to push image before final failure

### PR DESCRIPTION
**Pull request by TomSweeneyRedHat**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

Add the transport "docker://" to the destination image if it fails and there's no transport specified.  Then retry the lookup and finally fail if still not found.  Also renamed the variable DefaultRegistry to DefaultTransport and replaced a few "docker://" in the code with the variable.

### CODE REVIEW

**Comment by mheon** [comment]
@TomSweeneyRedHat We have a default transport in runtime.go (RuntimeConfig.ImageDefaultTransport) right now that can be set by the user, but it's not hooked into anything right now - can you either use that instead, or use that to set the value for DefaultTransport in runtime_img.go?

**Comment by TomSweeneyRedHat** [comment]
@mheon I'd forgotten about the RuntimeConfig.ImageDefaultTransport .  Will see what I can do with it, thanks!

File: libpod/runtime_img.go [hunk 1]
```diff
@@ -33,9 +33,9 @@ import (
 // Runtime API
 
 const (
-	// DefaultRegistry is a prefix that we apply to an image name
+	// DefaultTransport is a prefix that we apply to an image name
 	// to check docker hub first for the image
-	DefaultRegistry = "docker://"
+	DefaultTransport = "docker://"
```
Comment (baude) [inline libpod/runtime_img.go]: As long as you are at it, line 54ish in runtime.go has a struct that would benefit from this change too.

File: libpod/runtime_img.go [hunk 2]
```diff
@@ -613,7 +613,17 @@ func (r *Runtime) PushImage(source string, destination string, options CopyOptio
 	// Get the destination Image Reference
 	dest, err := alltransports.ParseImageName(destination)
 	if err != nil {
-		return errors.Wrapf(err, "error getting destination imageReference for %q", destination)
+		// add the docker:// transport to see if they neglected it.
+		if strings.Contains(destination, "://") {
```
Comment (baude) [inline libpod/runtime_img.go]: Depending on how you want to handle this @TomSweeneyRedHat , there is a slice for transports and we have a util.StringInSlice which takes a string and slice and returns a book depending on whether the string can be found in slice.

Comment (baude) [reply libpod/runtime_img.go]: or maybe it is time we break down and make an image-related function that is something like hasTransport ?

Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: I've made hasTransport, the stringinslice didn't fit as I've only a simple string at this point.

File: libpod/runtime_img.go [hunk 3]
```diff
@@ -219,7 +206,7 @@ func (k *Image) Decompose() error {
 		return nil
 	}
 	k.beenDecomposed = true
-	k.Transport = "docker://"
+	k.Transport = k.runtime.config.ImageDefaultTransport
```
Comment (TomSweeneyRedHat) [inline libpod/runtime_img.go]: Ugh, looks like I've a merge mess.  Will clean up tomorrow.

Comment (TomSweeneyRedHat) [reply libpod/runtime_img.go]: fixed the merge mess.

File: libpod/util.go [hunk 4]
```diff
@@ -35,10 +34,10 @@ func StringInSlice(s string, sl []string) bool {
 	return false
 }
 
-// FuncTimer helps measure the execution time of a function
-// For debug purposes, do not leave in code
-// used like defer FuncTimer("foo")
-func FuncTimer(funcName string) {
-	elapsed := time.Since(time.Now())
-	fmt.Printf("%s executed in %d ms\n", funcName, elapsed)
+// HasTransport determines if the image string contains '://', returns bool
```
Comment (rhatdan) [inline libpod/util.go]: Why are you removing FuncTimer?

Comment (TomSweeneyRedHat) [reply libpod/util.go]: Apparently my merge dance wasn't as smooth as it should have been.  I didn't men to remove that, back to dancing.

**Comment by TomSweeneyRedHat** [comment]
@mheon, thought I saw a comment about HasTransport being global.  I can't find the comment, but have changed that in the code.  You're probably right, I don't think it will be needed outside of the libpod.

**Comment by mheon** [comment]
LGTM

**Comment by TomSweeneyRedHat** [comment]
Test failures aren't lining up, going to close and reopen to force a test.

**Comment by rhatdan** [comment]
@TomSweeneyRedHat gofmt errors.

**Comment by TomSweeneyRedHat** [comment]
This PR hates me, looks like I've run into Travis environment issues.

**Comment by TomSweeneyRedHat** [comment]
OK, going to close and reopen to retest.  Picked up a 503 on an unrelated test.

**Comment by TomSweeneyRedHat** [comment]
Finally, all happy green buttons!

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
@rh-atomic-bot retry

**Comment by mheon** [comment]
I think the CI PR that merged last night means this is no longer passing - @TomSweeneyRedHat can you rebase and repush to retrigger CI and see what's going on?

**Comment by TomSweeneyRedHat** [comment]
@mheon will do and arg, this is a real problem child PR!

**Comment by TomSweeneyRedHat** [comment]
For some reason Red Hat testing didn't like the if/else in hasTransports, have changed it to just a straight return.  It passed prior, perhaps new testing in place for the builds?

**Comment by TomSweeneyRedHat** [comment]
@rhatdan my problem child PR is finally green, can we merge this?

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Comment by rhatdan** [comment]
@rh-atomic-bot retry

**Pull request closed without merging**